### PR TITLE
Created an integration to use llms built for Apple's MLX library

### PR DESCRIPTION
**Pull request by dwight-foster**
# Description

I created an integration to use llms built for or converted to Apples MLX library. I use my mac often to do machine learning prototyping and wanted to be able to use llama index tools with llms that are optimized to run on it. It uses mlx-lm specifically. I referenced the anthropic and huggingface llm integrations to create this one.

Fixes # (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/llms/llama-index-llms-mlx/README.md
Comment (nerdai) [inline llama-index-integrations/llms/llama-index-llms-mlx/README.md:1]: Please do write a detailed README for this integration. This appears in our llamahub.ai where a lot of our users will discover our integration packages.

File: llama-index-integrations/llms/llama-index-llms-mlx/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+python_version = "3.8"
+
+[tool.poetry]
+authors = ["Dwight Foster dwightfoster03@gmail.com"]
+description = "llama-index llms mlx integration"
+exclude = ["**/BUILD"]
+license = "MIT"
+name = "llama-index-llms-mlx"
+readme = "README.md"
+version = "0.1.11"
```
Comment (nerdai) [inline llama-index-integrations/llms/llama-index-llms-mlx/pyproject.toml:33]: nit: should we start this at `0.1.0`?

File: llama-index-integrations/llms/llama-index-llms-mlx/tests/test_llms_mlx.py
Comment (nerdai) [inline llama-index-integrations/llms/llama-index-llms-mlx/tests/test_llms_mlx.py:1]: please add base test classes, similar to this one found here: https://github.com/run-llama/llama_index/blob/main/llama-index-integrations/llms/llama-index-llms-anthropic/tests/test_llms_anthropic.py

**Review by nerdai — COMMENTED** [review]
Thanks @dwight-foster! Taking an initial first pass on it. Looks like we need to tailor pants BUILD files as well.

Could you run `pants tailor ::` in the root of this project and commit/push changes?

If you don't have `pants` installed, then you can alternatively provide me access to your fork and I can commit changes to your fork/main-branch.

**Comment by dwight-foster** [comment]
Thank you for the help @nerdai. I am not really sure why it is still failing. The error log does not give a clear reason. Can you infer anything from it. This is my first time contributing a large integration to this repo so sorry if any of this stuff is obvious.

File: docs/BUILD [hunk 2]
```diff
@@ -1 +1,9 @@
 python_sources()
+
+poetry_requirements(
```
Comment (ovuruska) [inline docs/BUILD:3]: Hi @dwight-foster, 
Could you share some insights into why this change is required?

File: llama-index-integrations/llms/llama-index-llms-mlx/llama_index/llms/mlx/base.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+)
+from llama_index.core.base.llms.generic_utils import (
+    messages_to_prompt as generic_messages_to_prompt,
+)
+from llama_index.core.prompts.base import PromptTemplate
+from llama_index.core.types import BaseOutputParser, PydanticProgramMode
+
+import logging
+
+from typing import Any, Callable, Optional, Union
```
Comment (ovuruska) [inline llama-index-integrations/llms/llama-index-llms-mlx/llama_index/llms/mlx/base.py:35]: This import can be extended with `Sequence`. `Sequence` was imported in the first line.

**Review by ovuruska — COMMENTED** [review]
Hi @dwight-foster, I left two comments.

File: llama-index-integrations/llms/llama-index-llms-mlx/llama_index/llms/mlx/base.py (hunk 3 above)
Comment (dwight-foster) [reply llama-index-integrations/llms/llama-index-llms-mlx/llama_index/llms/mlx/base.py:35]: Ok

File: docs/BUILD (hunk 2 above)
Comment (dwight-foster) [reply docs/BUILD:3]: It was not required. I was trying to figure out why the build failed the first time and so I tried a bunch of ways at the same time. This was what other build files in llm integrations looked like so I changed it to that.

**Comment by dwight-foster** [comment]
My tests locally are succeeding. I have run pants test :: and pytest. If anyone has any recommendations please let me know. Any help is greatly appreciated.

**Comment by logan-markewich** [comment]
@dwight-foster maybe it was overkill, but since this package depends on specific hardware, I don't think it makes sense to run tests for this in our CICD

Since the existing test wasn't really doing anything, I just deleted it for now

**Comment by dwight-foster** [comment]
Ok thank you

**Review by logan-markewich — APPROVED** [review]
